### ISSUE TITLE
SC2154 allows variables to be assigned out of order

### ISSUE THREAD

**Issue by jonhiggs** [koalaman/shellcheck#2291]
#### For bugs
- SC2154
- shellcheck 0.7.2
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### Here's a snippet or screenshot that shows the problem:

```sh

#!/usr/bin/env bash
value=$((x+y))
x=1
y=2

echo "value is ${value}"

```

#### Here's what shellcheck currently says:

```
```

#### Here's what I wanted or expected to see:

```
value=$((x+y))
         ^-- SC2154: x is referenced but not assigned.
           ^-- SC2154: y is referenced but not assigned.
```

**Comment by koalaman** [comment]
You're right, of course. This is a known deficiency in ShellCheck's oversimplified dataflow analysis. It will require some substantial work to improve, so it's unlikely to be fixed in the short term.
